kv/kvserver/concurrency: TestConcurrencyManagerBasic failed

kv/kvserver/concurrency.TestConcurrencyManagerBasic [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2994818&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2994818&tab=artifacts#/) on master @ [64dad5f195e910196ce6c472b6cd02d4c260317c](https://github.com/cockroachdb/cockroach/commits/64dad5f195e910196ce6c472b6cd02d4c260317c):


```
=== RUN   TestConcurrencyManagerBasic
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestConcurrencyManagerBasic449174360
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestConcurrencyManagerBasic
    concurrency_manager_test.go:499: -- test log scope end --
--- FAIL: TestConcurrencyManagerBasic (4.97s)
=== RUN   TestConcurrencyManagerBasic/wait_self
    concurrency_manager_test.go:91: 
        testdata/concurrency_manager/wait_self:123: 
        output didn't match expected:
        @@ -1,7 +1,8 @@
         [-] finish reqTxnMiddle: finishing request
         [4] sequence reqTxn2: lock wait-queue event: done waiting
         [4] sequence reqTxn2: conflicted with 00000003-0000-0000-0000-000000000000 on "k" for 1.234s
         [4] sequence reqTxn2: acquiring latches
        +[4] sequence reqTxn2: waiting to acquire write latch k@20.000000000,1, held by write latch k@10.000000000,1
         [4] sequence reqTxn2: scanning lock table for conflicting locks
         [4] sequence reqTxn2: sequencing complete, returned guard
         
    --- FAIL: TestConcurrencyManagerBasic/wait_self (0.18s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestConcurrencyManagerBasic PKG=./pkg/kv/kvserver/concurrency TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/kv nvanbenschoten
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestConcurrencyManagerBasic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver/concurrency.TestConcurrencyManagerBasic [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3001776&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3001776&tab=artifacts#/) on master @ [bf71898e043692cc87d2c232fe3066054275ec29](https://github.com/cockroachdb/cockroach/commits/bf71898e043692cc87d2c232fe3066054275ec29):


```
=== RUN   TestConcurrencyManagerBasic
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestConcurrencyManagerBasic375677879
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestConcurrencyManagerBasic
    concurrency_manager_test.go:517: -- test log scope end --
--- FAIL: TestConcurrencyManagerBasic (6.27s)
=== RUN   TestConcurrencyManagerBasic/deadlocks/lock_ordering_complex
    datadriven.go:243: 
        testdata/concurrency_manager/deadlocks:406: 
        output didn't match expected:
        @@ -1,7 +1,8 @@
         [-] finish req4w: finishing request
         [7] sequence req3w2: lock wait-queue event: done waiting
         [7] sequence req3w2: conflicted with 00000004-0000-0000-0000-000000000000 on "a" for 1.234s
         [7] sequence req3w2: acquiring latches
        +[7] sequence req3w2: waiting to acquire write latch a@10.000000000,1, held by write latch a@10.000000000,1
         [7] sequence req3w2: scanning lock table for conflicting locks
         [7] sequence req3w2: sequencing complete, returned guard
         
        --- FAIL: TestConcurrencyManagerBasic/deadlocks/lock_ordering_complex (0.23s)
=== RUN   TestConcurrencyManagerBasic/deadlocks
    --- FAIL: TestConcurrencyManagerBasic/deadlocks (0.47s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestConcurrencyManagerBasic PKG=./pkg/kv/kvserver/concurrency TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/kv nvanbenschoten
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestConcurrencyManagerBasic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver/concurrency.TestConcurrencyManagerBasic [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3037807&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3037807&tab=artifacts#/) on master @ [f5b21cf904be4a6273aefe573b70fb43f1f5ee3a](https://github.com/cockroachdb/cockroach/commits/f5b21cf904be4a6273aefe573b70fb43f1f5ee3a):


```
=== RUN   TestConcurrencyManagerBasic
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestConcurrencyManagerBasic640662966
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestConcurrencyManagerBasic
    concurrency_manager_test.go:517: -- test log scope end --
--- FAIL: TestConcurrencyManagerBasic (6.42s)
=== RUN   TestConcurrencyManagerBasic/deadlocks/lock_ordering_complex
    datadriven.go:243: 
        testdata/concurrency_manager/deadlocks:406: 
        output didn't match expected:
        @@ -1,7 +1,8 @@
         [-] finish req4w: finishing request
         [7] sequence req3w2: lock wait-queue event: done waiting
         [7] sequence req3w2: conflicted with 00000004-0000-0000-0000-000000000000 on "a" for 1.234s
         [7] sequence req3w2: acquiring latches
        +[7] sequence req3w2: waiting to acquire write latch a@10.000000000,1, held by write latch a@10.000000000,1
         [7] sequence req3w2: scanning lock table for conflicting locks
         [7] sequence req3w2: sequencing complete, returned guard
         
        --- FAIL: TestConcurrencyManagerBasic/deadlocks/lock_ordering_complex (0.27s)
=== RUN   TestConcurrencyManagerBasic/deadlocks
    --- FAIL: TestConcurrencyManagerBasic/deadlocks (0.54s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestConcurrencyManagerBasic PKG=./pkg/kv/kvserver/concurrency TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/kv nvanbenschoten
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestConcurrencyManagerBasic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/concurrency/concurrency_manager.go
+++ b/pkg/kv/kvserver/concurrency/concurrency_manager.go
@@ -296,11 +296,27 @@ func shouldAcquireLatches(req Request) bool {
 
 // FinishReq implements the RequestSequencer interface.
 func (m *managerImpl) FinishReq(g *Guard) {
-	if ltg := g.moveLockTableGuard(); ltg != nil {
-		m.lt.Dequeue(ltg)
-	}
+	// NOTE: we release latches _before_ exiting lock wait-queues deliberately.
+	// Either order would be correct, but the order here avoids non-determinism in
+	// cases where a request A holds both latches and lock wait-queue reservations
+	// and has a request B waiting on its reservations. If request A released its
+	// reservations before releasing its latches, it would be possible for B to
+	// beat A to the latch manager and end up blocking on its latches briefly. Not
+	// only is this confusing in traces, but it is slightly less efficient than if
+	// request A released latches before letting anyone waiting on it in the lock
+	// table proceed, ensuring that waiters do not hit its latches.
+	//
+	// Elsewhere, we relate the relationship of between the latch manager and the
+	// lock-table to that of a mutex and condition variable pair. Following that
+	// analogy, this release ordering is akin to signaling a condition variable
+	// after releasing its associated mutex. Doing so ensures that whoever the
+	// signaler wakes up (if anyone) will never bump into its mutex immediately
+	// upon resumption.
 	if lg := g.moveLatchGuard(); lg != nil {
 		m.lm.Release(lg)
+	}
+	if ltg := g.moveLockTableGuard(); ltg != nil {
+		m.lt.Dequeue(ltg)
 	}
 	releaseGuard(g)
 }

